serde/src/main/java/com/amazon/ionhiveserde/formats/IonOutputFormat.java

```diff
@@ -76,15 +80,27 @@
                                                              final Progressable progress)
         throws IOException {
 
-        final FileSystem fs = finalOutPath.getFileSystem(jc);
-        final OutputStream out = fs.create(finalOutPath, progress);
-
-        // If we are passed in a reporter, make sure we call incrCounters during write
-        Optional<Reporter> reporter = Optional.empty();
-        if (progress instanceof Reporter) {
-            reporter = Optional.of((Reporter) progress);
+        // I believe we should expect the following to hold:
+        // isCompressed == FileOutputFormat.getCompressOutput(jc);
+        // A layer farther up should be getting this flag from the job configuration
+        if (isCompressed) {
+            CompressionCodec codec = getCompressionCodec(jc);
+            Path file = FileOutputFormat.getTaskOutputPath(jc,
+                    finalOutPath.getName() + codec.getDefaultExtension());
+            FileSystem fs = file.getFileSystem(jc);
+            FSDataOutputStream fileOut = fs.create(file, progress);
+            return new IonRecordWriter(new DataOutputStream(codec.createOutputStream(fileOut)), null);
```

Comment:
Just a quick note: we'll need to add a reporter if the progress object is of the correct type in order to update the BytesWrittenCounter properly.

Reply:
Let me looking into on it.

One minor thing: I used `return new IonRecordWriter(new DataOutputStream(codec.createOutputStream(fileOut)), Optional.empty());` instead of null type for reporter in the latest commit.

Reply:
Yup! I can submit a followup if we just want to do the tweak after this PR gets merged.

Reply:
Following up on our slack discussion:

Instead of storing OutputStream in the RecordWriter, we can use DataOutputStream. So something like this:

```
// If we are passed in a reporter, make sure we call incrCounters during write
        Optional<Reporter> reporter = Optional.empty();
        if (progress instanceof Reporter) {
            reporter = Optional.of((Reporter) progress);
        }
        if (isCompressed) {
            CompressionCodec codec = getCompressionCodec(jc);
            FileSystem fs = finalOutPath.getFileSystem(jc);
            FSDataOutputStream fileOut = fs.create(finalOutPath, progress);
            DataOutputStream compressedStream = new DataOutputStream(codec.createOutputStream(fileOut));
            return new IonRecordWriter(compressedStream, reporter);
        } else {
            final FileSystem fs = finalOutPath.getFileSystem(jc);
            final FSDataOutputStream out = fs.create(finalOutPath, progress);
            return new IonRecordWriter(out, reporter);
        }
```

Then during write we can grab the size of the output stream before the write:
```
long pos = out.size();
```

and call updateBytesWritten with the difference after the write:
```
out.write(bytesWritable.getBytes(), 0, bytesWritten);
updateBytesWritten(out.size() - pos);
```

Reply:
Below commit handle compressed file bytes size.

As discussed offline: 
> Some compression codecs like zstd build up a list of input and output patterns so we cant actually get the compressed output size until after the compression occurs and the buffer is flushed